src/ratelimit_manager.cpp

```diff
@@ -0,0 +1,588 @@
+            } else {
+                // API key is banned
+                return true;
+            }
+        }
+
+        if (!api_key_request_counts.contains(api_key)) {
+            api_key_request_counts.emplace(api_key, request_counter_t {});
+        }
+
+        // Check last reset time for minute rate limit 
+        if (api_key_request_counts[api_key].last_reset_time_minute + 60 < std::chrono::system_clock::to_time_t(std::chrono::system_clock::now())) {
+            // Reset minute rate limit
+            api_key_request_counts[api_key].previous_requests_count_minute = api_key_request_counts[api_key].current_requests_count_minute;
+            api_key_request_counts[api_key].current_requests_count_minute = 0;
+            api_key_request_counts[api_key].last_reset_time_minute = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+        }
+
+        // Check last reset time for hour rate limit 
+        if (api_key_request_counts[api_key].last_reset_time_hour + 3600 < std::chrono::system_clock::to_time_t(std::chrono::system_clock::now())) {
+            // Reset hour rate limit
+            api_key_request_counts[api_key].previous_requests_count_hour = api_key_request_counts[api_key].current_requests_count_hour;
+            api_key_request_counts[api_key].current_requests_count_hour = 0;
+            api_key_request_counts[api_key].last_reset_time_hour = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+        }
+
+        // Increase current request count for minute rate limit and hour rate limit
+        api_key_request_counts[api_key].current_requests_count_minute++;
+        api_key_request_counts[api_key].current_requests_count_hour++;
+
+        time_t currentTime;
+        struct tm * timeinfo;
+        std::time( & currentTime);
+        timeinfo = std::localtime( & currentTime);
+
+        // Check if current request count for minute rate limit is over the limit with sliding window
+        auto current_rate_minute = (60 - (timeinfo -> tm_sec)) / 60.0 * api_key_request_counts[api_key].previous_requests_count_minute + api_key_request_counts[api_key].current_requests_count_minute;
+
+        if (api_key_rate_limits[api_key].minute_rate_limit >= 0 && current_rate_minute >= api_key_rate_limits[api_key].minute_rate_limit) {
+            ban_api_key_wrapped(api_key);
+            return true;
+        }
+
+        // Check if current request count for hour rate limit is over the limit with sliding window
+        auto current_rate_hour = (3600 - (timeinfo -> tm_min * 60 + timeinfo -> tm_sec)) / 3600.0 * api_key_request_counts[api_key].previous_requests_count_hour + api_key_request_counts[api_key].current_requests_count_hour;
+        if (api_key_rate_limits[api_key].hour_rate_limit >= 0 && current_rate_hour >= api_key_rate_limits[api_key].hour_rate_limit) {
+            ban_api_key_wrapped(api_key);
+            return true;
+        }
+    }
+
+    // Check if IP is tracked
```

Comment:
I see a lot of duplication in logic with the API key checks made above. This is making me think of just treating both of them as a singular entity. We can have a single hashmap, where the tracking identifier can can either be a IP address or an API key. The internal logic should not care what identifier is it. If we want we can prefix the key with "ip_" or "key_" to differentiate the entities. I feel that this will _vastly_ reduce the number of data structures we need to maintain and remove duplication of logic by making everything generic. What do you think?